src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -546,6 +555,67 @@ messageInstD ctxt parentIdent msgIdent messageParts =
                       , decodeMessageDecl
                       , dotProtoDecl ]))
 
+toJSONPBMessageInstD :: TypeContext -> DotProtoIdentifier -> DotProtoIdentifier
+                     -> [DotProtoMessagePart] -> CompileResult HsDecl
+toJSONPBMessageInstD _ctxt parentIdent msgIdent messageParts =
+  do
+  msgName <- nestedTypeName parentIdent =<< dpIdentUnqualName msgIdent
+  kvps    <- sequence
+               [ (,"f" ++ show n) <$> dpIdentUnqualName fldIdent
+               | DotProtoMessageField
+                   (DotProtoField (FieldNumber n) _ fldIdent _ _)
+                   <- messageParts
+               ]
+  let toEncodingPBDecl =
+        match_ (HsIdent "toEncodingPB")
+               [ patVar "opts"
+               , HsPApp (unqual_ msgName) [ patVar varNm | (_, varNm) <- kvps ]
+               ]
+               (HsUnGuardedRhs toEncodingPBE) []
+        where
+          toEncodingPBE =
+            apply (HsVar (jsonpbName "fieldsPB"))
+                  [ HsVar (unqual_ "opts")
+                  , HsList [ HsInfixApp (HsLit (HsString fldNm))
+                                        toJSONPBOp
+                                        (HsVar (unqual_ varNm))
+                           | (fldNm, varNm) <- kvps
+                           ]
+                  ]
+  pure (instDecl_ (jsonpbName "ToJSONPB")
+                 [ type_ msgName ]
+                 [ HsFunBind [ toEncodingPBDecl ] ])
+
+fromJSONPBMessageInstD :: TypeContext -> DotProtoIdentifier -> DotProtoIdentifier
+                       -> [DotProtoMessagePart] -> CompileResult HsDecl
+fromJSONPBMessageInstD _ctxt parentIdent msgIdent messageParts =
+  do
+  msgName <- nestedTypeName parentIdent =<< dpIdentUnqualName msgIdent
+  kvps    <- sequence
```

Comment:
Since this same code appears in `toJSONPBMessageInstD` I think factoring it out into a separate function is a good idea. This is a nit though so it's non-critical.